crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+            _ => value_ty,
+        };
+
+        self.add_binding(variable.into(), definition, target_ty);
         self.types
             .expressions
-            .insert(target.scoped_ast_id(self.db, self.scope), value_ty);
+            .insert(variable.scoped_ast_id(self.db, self.scope), target_ty);
+    }
+
+    fn infer_sequence_unpacking(
+        &mut self,
+        target: &ast::Expr,
+        value_ty: Type<'db>,
+        variable: &ast::ExprName,
+    ) -> Type<'db> {
+        // The inner function is recursive and only differs in the return type which is an `Option`
+        // where if the variable is found, the corresponding type is returned otherwise `None`.
+        fn inner<'db>(
+            builder: &mut TypeInferenceBuilder<'db>,
+            target: &ast::Expr,
+            value_ty: Type<'db>,
+            variable: &ast::ExprName,
+        ) -> Option<Type<'db>> {
+            match target {
+                // TODO: Handle other possible target expression like attribute, subscript,
+                // and starred.
+                ast::Expr::Name(name) if name == variable => {
+                    return Some(value_ty);
+                }
+                ast::Expr::List(ast::ExprList { elts, .. })
+                | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => match value_ty {
+                    Type::Tuple(tuple_ty) => {
+                        let starred_index = elts.iter().position(ast::Expr::is_starred_expr);
+
+                        let element_types = if let Some(starred_index) = starred_index {
+                            if tuple_ty.len(builder.db) >= elts.len() {
+                                let mut element_types = Vec::with_capacity(elts.len());
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty.elements(builder.db).get(..starred_index).unwrap(),
+                                );
+
+                                // E.g., in `(a, *b, c, d) = ...`, the index of starred element `b`
+                                // is 1 and the remaining elements after that are 2.
+                                let remaining = elts.len() - 2;
+                                let end_index =
+                                    tuple_ty.len(builder.db) - starred_index - remaining;
+                                // SAFETY: Safe because of the length check above.
+                                let _starred_element_types = tuple_ty
+                                    .elements(builder.db)
+                                    .get(starred_index..=end_index)
+                                    .unwrap();
+                                // TODO: Combine the types into a list type.
+                                // combine_types(starred_element_types);
+                                element_types.push(Type::Todo);
```

Comment:
Somewhere this Todo gets swallowed and turned into Unknown. Not sure why, and it's not really a big deal, but if it's not hard to preserve it as a Todo that would be nice.

Reply:
I got it, it's because the `inner` function does not account for `Expr::Starred` and only checks for `Expr::Name`. So, it would return `None` which then defaults to `Unknown`.

So, the diff would become
```diff
                 ast::Expr::Name(name) if name == variable => {
                     return Some(value_ty);
                 }
+                ast::Expr::Starred(ast::ExprStarred { value, .. }) => {
+                    // TODO: Wrap the `value_ty` in a list type.
+                    return inner(builder, value, value_ty, variable);
+                }
                 ast::Expr::List(ast::ExprList { elts, .. })
                 | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => match value_ty {
                     Type::Tuple(tuple_ty) => {
                         let starred_index = elts.iter().position(ast::Expr::is_starred_expr);
```